lib/bitcrowd_ecto/changeset.ex

```diff
@@ -425,4 +425,118 @@ defmodule BitcrowdEcto.Changeset do
+  Introspects a schema and casts all defined fields from a params map.
+
+  - Accepts a schema module or structs or changesets.
+  - Can deal with embeds.
+
+  ## Options
+
+  - `required` list of required field names
+  - `optional` list of optional field names (= inverse set is required)
+
+  `required` and `optional` options must not be present at the same time.
+
+  ## Examples
+
+      iex> changeset = auto_cast(TestEmbeddedSchema, %{some_field: 4})
+      iex> changeset.valid?
+      true
+
+      iex> changeset = auto_cast(%TestEmbeddedSchema{}, %{some_field: 4})
+      iex> changeset.valid?
+      true
+
+      iex> changeset = auto_cast(change(%TestEmbeddedSchema{}), %{some_field: 4})
+      iex> changeset.valid?
+      true
+
+      iex> changeset = auto_cast(TestEmbeddedSchema, %{}, required: [:some_field])
+      iex> changeset.errors
+      [some_field: {"can't be blank", [validation: :required]}]
+
+      iex> changeset = auto_cast(TestEmbeddedSchema, %{}, optional: [:some_other_field])
+      iex> changeset.errors
+      [some_field: {"can't be blank", [validation: :required]}]
+
+  """
+  @doc since: "0.17.0"
+  @spec auto_cast(module | struct, map) :: Ecto.Changeset.t()
+  @spec auto_cast(module | struct, map, [auto_cast_option]) :: Ecto.Changeset.t()
+  def auto_cast(schema_or_struct_or_changeset, params, opts \\ [])
+
+  def auto_cast(%Ecto.Changeset{} = changeset, params, opts) do
+    do_auto_cast(changeset.data.__struct__, changeset, params, opts)
+  end
+
+  def auto_cast(schema, params, opts) when is_atom(schema) do
+    do_auto_cast(schema, struct!(schema), params, opts)
+  end
+
+  def auto_cast(struct, params, opts) when is_struct(struct) do
+    do_auto_cast(struct.__struct__, struct, params, opts)
+  end
+
+  defp do_auto_cast(schema, struct_or_changeset, params, opts) do
+    required = required_fields(schema, opts)
+    grouped = grouped_fields(schema)
+
+    struct_or_changeset
+    |> cast_scalars(params, grouped.scalars, required)
+    |> cast_embeds(grouped.embeds, required)
+  end
+
+  defp required_fields(schema, opts) do
+    required = Keyword.get(opts, :required)
+    optional = Keyword.get(opts, :optional)
+
+    cond do
+      required && optional ->
+        raise ArgumentError, ":required and :optional options are mutually exclusive"
+
+      required ->
+        required
+
+      optional ->
+        schema.__schema__(:fields) -- optional
+
+      true ->
+        []
+    end
+  end
+
+  defp grouped_fields(schema) do
+    :fields
+    |> schema.__schema__()
+    |> Enum.group_by(fn field ->
+      case schema.__schema__(:type, field) do
+        {:parameterized, Ecto.Embedded, _} ->
+          :embeds
+
+        # Simplification, can be extended as needed.
+        _other ->
+          :scalars
+      end
+    end)
+    |> Map.put_new(:embeds, [])
+    |> Map.put_new(:scalars, [])
+  end
+
+  defp cast_scalars(schema_struct, params, scalars, required) do
+    required = scalars -- scalars -- required
```

Comment:
`blocker`

How does this works? It would always result in an empty array...

```suggestion
    required = scalars -- scalars -- required
```

Reply:
:thinking: I'm confused, this had parenthesis before, must have been removed by the formatter?! it's probably wrong, let me check

Reply:
this is so weird. The expression is evaluated as intended, but the formatter removes the parenthesis, and apparently the `--` operator evaluates right-to-left?!?

```
iex(1)> [1, 2] -- [1] -- [2]
[2]
iex(2)> [1, 2] -- ([1] -- [2])
[2]
```

Reply:
indeed, it's even documented:

https://hexdocs.pm/elixir/1.12/operators.html#operator-precedence-and-associativity

![image](https://github.com/bitcrowd/bitcrowd_ecto/assets/96114/126e5590-1b0c-409a-bcf5-cc51b7f13d4d)

i'll add a comment

Reply:
🤯 

Then please add a comment for future readers.